### ISSUE TITLE
ZLayer succeed, fromEffect, and fromManaged inconsistent?

### ISSUE THREAD

**Issue by dariusrobson** [zio/zio#2929]
```ZLayer.fromEffect``` and ```ZLayer.fromManaged``` accept a IO/Managed where ```A <: Has[_]```
whereas ```ZLayer.succeed``` just accepts ```A``` (and the implementation wraps this in ```Has(a)```). 

Client code to create a Layer for effectful/managed resources requires wrapping in ```Has(_)``` but this is not the case for values. Is this inconsistent? (and should it be addressed?)

```scala
def fromEffect[E, A <: Has[_]](zio: IO[E, A]): ZLayer.NoDeps[E, A]

def fromManaged[E, A <: Has[_]](m: Managed[E, A]): ZLayer.NoDeps[E, A]

def succeed[A: Tagged](a: => A): ZLayer.NoDeps[Nothing, Has[A]]```

**Comment by dariusrobson** [comment]
@iravid @adamgfraser I have created an issue to discuss these ZLayer signatures (as mentioned on discord)

**Comment by iravid** [comment]
Just doing `A: Tagged` on all versions and not requiring `A <: Has[_]` seems more ergonomic to me. cc @adamgfraser

**Comment by adamgfraser** [comment]
I think that makes sense. Will update tonight.

**Comment by adamgfraser** [comment]
So looking at this more I actually don't think we can do this. It makes the simple case slightly more ergonomic but makes anything more complicated significantly more difficult. The issue is that `Has` lets us safely express multiple dependencies, whereas with `Tagged` we can only express one dependency at a time.

For example, consider the definition of the `TestConsole` service:

```scala
def live(data: Data): ZLayer.NoDeps[Nothing, Console with TestConsole] =
  ZLayer.fromEffect(
    Ref.make(data).map(ref => Has.allOf[Console.Service, TestConsole.Service](Test(ref), Test(ref)))
  )
```

Here the `live` layer outputs both the `Console` and `TestConsole` services. With `Has` this is no problem because we can use the `Has.allOf` constructor to create `Has[Console.Service] with Has[TestConsole.Service]`. In contrast, if we need an `A` value with `Tagged` we are stuck because to express both dependencies we would have to use an intersection type which is exactly what we have to avoid. More generally when we are working with more abstract service bundles all we know is that they are a subtype of `Has` and don't necessarily have tags for them. We could always create single service layers and combine them with `++`, but losing the ability to easily build layers with multiple services as a single layer seems like a significant cost for not much benefit.

We could potentially add some additional methods, `fromManagedService` and `fromEffectService` that would take a single service but I think those risk significant confusion with the `fromServicesM` and `fromServicesManaged` variants and just providing multiple ways to do the same thing.

I think if we did anything here I would prefer to align everything with requiring `Has` which requires slightly more typing in some situations but is significantly more compositional.

**Comment by iravid** [comment]
Ahh, that makes sense. Thank you for clarifying @adamgfraser. I am 👍 on changing `succeed` to `[A <: Has[_]]` and coming up later with good names for the `A: Tagged` versions of the other constructors. 

Agreed on the potential confusing on `fromServicesM/fromServicesManaged`, though I think those constructors are addressing usecases that are far less common than `fromManagedService/fromEffectService`.

**Comment by dariusrobson** [comment]
I think that having both (```A <: Has[_]``` and ```A: Tagged```) versions would be ideal.

I was doing something like this:

```scala
    def live(settings: SessionSettings): ZLayer.NoDeps[Nothing, Session] =
      ZLayer.fromManaged(UIO(DSession.build).map { underlying =>
        new Service {
          def close: UIO[Unit] =
            javaz.fromCompletionStage(underlying.closeAsync).unit.orDie
        }
      }.toManaged(_.close).map(Has(_)))
```

and expected a way to remove the ```.map(Has(_))``` boilerplate
But I completely see the problems with naming conflicts.

**Comment by adamgfraser** [comment]
Resolves by #2937.
